src/_C014.ino

```diff
@@ -0,0 +1,686 @@
+        pubname = CPLUGIN_014_BASE_TOPIC; // Scheme to form device messages
+        pubname.replace(F("%sysname%"), Settings.Name);
+        // disconnected: this is the state the device is in when it is cleanly disconnected from the MQTT broker. You must send this message before cleanly disconnecting
+        CPlugin_014_sendMQTTdevice(pubname,"$state","disconnected",errorCounter);
+        break;
+      }
+
+    case CPLUGIN_FALLING_ASLEEP:
+      {
+        pubname = CPLUGIN_014_BASE_TOPIC; // Scheme to form device messages
+        pubname.replace(F("%sysname%"), Settings.Name);
+        // sleeping: this is the state the device is in when the device is sleeping. You have to send this message before sleeping.
+        CPlugin_014_sendMQTTdevice(pubname,"$state","sleeping",errorCounter);
+        break;
+      }
+
+    case CPLUGIN_PROTOCOL_RECV:
+      {
+        byte ControllerID = findFirstEnabledControllerWithId(CPLUGIN_ID_014);
+        if (ControllerID == CONTROLLER_MAX) {
+          // Controller is not enabled.
+          break;
+        } else {
+          String cmd;
+          struct EventStruct TempEvent;
+          TempEvent.TaskIndex = event->TaskIndex;
+          TempEvent.Source = VALUE_SOURCE_MQTT; // to trigger the correct feedback
+          bool validTopic = false;
+          int lastindex = event->String1.lastIndexOf('/');
+          const String lastPartTopic = event->String1.substring(lastindex + 1);
+          errorCounter = 0;
+          log=F("C014 : MQTT received: ");
+          if (event->String1.substring(lastindex + 1) == F("set"))
+          {
+            pubname = event->String1.substring(0,lastindex);
+            lastindex = pubname.lastIndexOf('/');
+            String nodeName = pubname.substring(0,lastindex);
+            String valueName = pubname. substring(lastindex+1);
+            lastindex = nodeName.lastIndexOf('/');
+            nodeName = nodeName. substring(lastindex+1);
+            log+=F("/set: N: ");
+            log+=nodeName;
+            log+=F(" V: ");
+            log+=valueName;
+
+            if (nodeName == F(CPLUGIN_014_SYSTEM_DEVICE)) // msg to a system device
+            {
+              if (valueName.substring(0,strlen(CPLUGIN_014_GPIO_VALUE)) == F(CPLUGIN_014_GPIO_VALUE)) // msg to to set gpio values
+              {
+                cmd = ("GPIO,");
+                cmd += valueName.substring(strlen(CPLUGIN_014_GPIO_VALUE)).toInt(); // get the GPIO
+                if (event->String2=="true" || event->String2=="1") cmd += F(",1");
+                  else cmd += F(",0");
+                validTopic = true;
+              } else if (valueName==CPLUGIN_014_CMD_VALUE) // msg to send a command
+              {
+                cmd = event->String2;
+                validTopic = true;
+              } else
+              {
+                cmd = "SYSTEM not found";
+              }
+            } else if (nodeName == F(CPLUGIN_014_CMD_VALUE)) // legethy homie/system/cmd topic
```

Comment:
This kind of nested if... else scopes is very hard to read.
They do have the proper indentation, but still it would make sense to have a single level of if...else checks for the `nodeName` and in there call a separate function to handle those cases.

Also, what is "legethy" in the comment? "legacy" ?

Reply:
You are right. Thats when the code is growing ... Have to sort it. Was started with a quick modification of the original openHAB controller ended much bigger than expected. That's fine I'll clean it up.
As I used the original openHAB controller as a template and I was keeping as much code as possible. The "legacy" (code & comment) could and should be erased now.